File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        //distance from source to all other
        vector<int>dist(n,INT_MAX);
        dist[src]=0;
        //run only K=1 times since we want shortest distance in K hops
        for(int i=0;i<=k;i++){
            //create a copy of dist vector
            vector<int>temp(dist);
            // Iterate over all flights
            for (auto& flight : flights) {
                int u = flight[0], v = flight[1], w = flight[2];  // u = source, v = destination, w = cost
                
                // If the current distance to 'u' is not INT_MAX, update distance to 'v'
                if (dist[u] != INT_MAX) {
                    temp[v] = min(temp[v], dist[u] + w);
                }
            }
            dist=temp;
        }
        return dist[dst]==INT_MAX?-1:dist[dst];
    }
};

File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        HashMap<Integer,List<int[]>>map = new HashMap<>();
        for(int[] flight: flights){
            int from = flight[0], to = flight[1], price = flight[2];
            if(! map.containsKey(from)){
                map.put(from, new ArrayList<>());
            }
            map.get(from).add(new int[]{to, price});
        }
        Queue<int[]>queue = new LinkedList<>();
        queue.add(new int[]{0, src, 0});
        int[] dist = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        while(!queue.isEmpty()){
            int[] curCity = queue.poll();
            int city = curCity[1], price = curCity[0], stops = curCity[2];
            if(stops > k) continue;
            if(!map.containsKey(city)) continue;
            for(int[] next: map.get(city)){
                int nextCity = next[0], cost = price + next[1];
                if(dist[nextCity] > cost){
                    dist[nextCity] = cost;
                    queue.add(new int[]{cost, nextCity, stops +1});
                }
            }
        }
        return  dist[dst] == Integer.MAX_VALUE? -1 : dist[dst];
        
    }
}